File: algorithms/insertion_sort.c
#include <stdio.h>

void insertion_sort(int *a, size_t n) {
  for (size_t i = 1; i < n; i++) {
    int j = i - 1;
    int h = a[i];
    while (j >= 0 && h < a[j]) {
      a[j+1] = a[j];
      j = j - 1;
    }
    a[j+1] = h;
  }
}

int main(void) {
  size_t n = 5;
  int a[] = {6, 3, 5, 1, 9};

  for (size_t i = 0; i < n; i++) {
    printf("%d ", a[i]);
  }
  printf("\n---\n");

  insertion_sort(a, n);

  for (size_t i = 0; i < n; i++) {
    printf("%d ", a[i]);
  }
  printf("\n");

  return 0;
}

File: algorithms/insertion_sort.py
def insertion_sort(a: list):
  n = len(a)
  for i in range(1, n):
    h = a[i]
    j = i-1
    while a[j] > h and j >= 0:
      a[j+1] = a[j]
      j -= 1
    a[j+1] = h

if __name__ == "__main__":
  arr = [5, 2, 4, 6, 1, 3]

  print("before:", arr)
  insertion_sort(arr)
  print("after:", arr)
